flink-contrib/flink-timerserivce-rocksdb/src/main/java/org/apache/flink/contrib/streaming/api/operators/RocksDBInternalTimerService.java

```diff
@@ -0,0 +1,797 @@
+
+		keyGroup -= keyGroupRange.getStartKeyGroup();
+
+		if (keyGroup >= (numKeyGroupsPerPartition + 1L) * numFatPartitions) {
+			return (int)((keyGroup - (numKeyGroupsPerPartition + 1L) * numFatPartitions) / numKeyGroupsPerPartition + numFatPartitions);
+		} else {
+			return (int)(keyGroup / (numKeyGroupsPerPartition + 1L));
+		}
+	}
+
+	/**
+	 * Compute the range of the given partition
+	 */
+	private static KeyGroupRange getRangeForPartition(KeyGroupRange keyGroupRange, int partitionIndex, int numPartitions) {
+		Preconditions.checkArgument(keyGroupRange != null, "The range must not be null");
+		Preconditions.checkArgument(partitionIndex >= 0, "Partition index must be not smaller than zero.");
+		Preconditions.checkArgument(numPartitions > 0, "Partition count must be greater than zero.");
+
+		long numKeysPerPartition = (keyGroupRange.getEndKeyGroup() - keyGroupRange.getStartKeyGroup() + 1L) / numPartitions;
+		long numFatPartitions = (keyGroupRange.getEndKeyGroup() - keyGroupRange.getStartKeyGroup() + 1L) - numKeysPerPartition * numPartitions;
+
+		if (partitionIndex >= numFatPartitions) {
+			int startKeyGroup = keyGroupRange.getStartKeyGroup() + (int)(numFatPartitions * (numKeysPerPartition + 1L) + (partitionIndex - numFatPartitions) * numKeysPerPartition);
+			int endKeyGroup = (int)(startKeyGroup + numKeysPerPartition - 1L);
+
+			return (startKeyGroup > endKeyGroup ? null : new KeyGroupRange(startKeyGroup, endKeyGroup));
+		} else {
+			int startKeyGroup = keyGroupRange.getStartKeyGroup() + (int)(partitionIndex * (numKeysPerPartition + 1L));
+			int endKeyGroup = (int)(startKeyGroup + numKeysPerPartition);
+
+			return new KeyGroupRange(startKeyGroup, endKeyGroup);
+		}
+	}
+
+	// ------------------------------------------------------------------------
+	//  Serialization Methods
+	// ------------------------------------------------------------------------
+
+	private byte[] serializeKeyGroup(int keyGroup) {
+		ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+		DataOutputViewStreamWrapper outputView = new DataOutputViewStreamWrapper(outputStream);
+
+		try {
+			IntSerializer.INSTANCE.serialize(keyGroup, outputView);
+		} catch (IOException e) {
+			throw new RuntimeException("Error while deserializing the key group.", e);
+		}
+
+		return outputStream.toByteArray();
+	}
+
+	private byte[] serializeRawTimer(Tuple4<Integer, Long, K, N> rawTimer) {
+		ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+		DataOutputViewStreamWrapper outputView = new DataOutputViewStreamWrapper(outputStream);
+
+		try {
+			IntSerializer.INSTANCE.serialize(rawTimer.f0, outputView);
+			LongSerializer.INSTANCE.serialize(rawTimer.f1, outputView);
+			keySerializer.serialize(rawTimer.f2, outputView);
+			namespaceSerializer.serialize(rawTimer.f3, outputView);
+		} catch (IOException e) {
+			throw new RuntimeException("Error while serializing the raw timer.", e);
```

Comment:
I would disencourage the usage of plain `RuntimeException` for the same reason we use more specific subclasses of `Exception`, like  `IOException`. Better use a more specific subclass of `RuntimeException`, like `IllegalStateException` or whatever looks more appropriate. This should be changed in serveral places in this class.